### PR TITLE
Program address generator

### PR DESCRIPTION
**Pull request by jackcmay**
#### Problem

No way to generate program addresses used by programs to sign cross-program invocations

#### Summary of Changes

Provide one adherent to https://github.com/solana-labs/solana/pull/8155

These changes are part of a larger effort that has been broken into logical pieces to ease review

Fixes #

### CODE REVIEW

File: sdk/src/pubkey.rs [hunk 1]
```diff
@@ -79,6 +79,18 @@ impl Pubkey {
         ))
     }
 
+    pub fn create_program_address(
+        seeds: &[&str],
+        program_id: &Pubkey,
+    ) -> Result<Pubkey, PubkeyError> {
+        // Use a deterministic base for all program addresses that are owned by `program_id`
+        let mut base = Pubkey::create_with_seed(program_id, &"ProgramDerivedAddress", program_id)?;
```
Comment (aeyakovenko) [inline sdk/src/pubkey.rs]: What does this function do?

Comment (aeyakovenko) [reply sdk/src/pubkey.rs]: I think we should use hash directly, since it’s going to have impact on security.  Hash all the strings then haha the output.

Comment (aeyakovenko) [reply sdk/src/pubkey.rs]: I think we should use hash directly, since it’s going to have impact on security.  Hash all the strings then hash the output.

Comment (jackcmay) [reply sdk/src/pubkey.rs]: @aeyakovenko Is this more whast you had in mind?

File: sdk/src/pubkey.rs [hunk 2]
```diff
@@ -79,6 +79,21 @@ impl Pubkey {
         ))
     }
 
+    pub fn create_program_address(
+        seeds: &[&str],
+        program_id: &Pubkey,
+    ) -> Result<Pubkey, PubkeyError> {
+        // Use a deterministic base for all program addresses that are owned by `program_id`
+        let mut hash = hashv(&["ProgramDerivedAddress".as_ref(), program_id.as_ref()]);
```
Comment (aeyakovenko) [inline sdk/src/pubkey.rs]: Use a Hash object, this creates a new one and closes it every time without additional benefit to security.  Our strings should be last.

Comment (jackcmay) [reply sdk/src/pubkey.rs]: @aeyakovenko 👀

**Pull request merged**